activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb

```diff
@@ -9,9 +9,9 @@ module ConnectionAdapters #:nodoc:
     # are typically created by methods in TableDefinition, and added to the
     # +columns+ attribute of said TableDefinition object, in order to be used
     # for generating a number of table creation or table changing SQL statements.
-    ColumnDefinition = Struct.new(:name, :type, :limit, :precision, :scale, :default, :null, :first, :after, :auto_increment, :primary_key, :collation, :sql_type, :comment, :as) do # :nodoc:
-      def primary_key?
-        primary_key || type.to_sym == :primary_key
+    ColumnDefinition = Struct.new(:name, :type, :options, :sql_type) do # :nodoc:
+      def [](key)
+        options[key]
```

Comment:
After reading the rest of the PR, perhaps we could provide these reader/writer methods without deprecation.